hnn_core/tests/test_extracellular.py

```diff
@@ -0,0 +1,338 @@
+
+    assert isinstance(joblib_net.rec_array['arr1'].get_data(), np.ndarray)
+    assert_array_equal(joblib_net.rec_array['arr1'].get_data().shape,
+                       [len(joblib_net.rec_array['arr1']._data),
+                        len(joblib_net.rec_array['arr1']._data[0]),
+                        len(joblib_net.rec_array['arr1']._data[0][0])])
+
+    # make sure sampling rate is fixed (raises RuntimeError if not)
+    _ = joblib_net.rec_array['arr1'].sfreq
+    # check plotting works
+    joblib_net.rec_array['arr1'].plot(show=False)
+
+
+def _mathematical_dipole(e_pos, d_pos, d_Q):
+    rr = e_pos - d_pos
+    R = norm(rr)
+    Q = norm(d_Q)
+    cosT = np.dot(rr, d_Q) / (R * Q)
+    return (Q * cosT) / (4 * np.pi * R ** 2)
+
+
+# require MPI to speed up due to large number of extracellular electrodes
+@requires_mpi4py
+def test_dipolar_far_field():
+    """Test that LFP in the far field is dipolar when expected."""
+    hnn_core_root = op.dirname(hnn_core.__file__)
+    params_fname = op.join(hnn_core_root, 'param', 'default.json')
+    params = read_params(params_fname)
+    params.update({'N_pyr_x': 3,
+                   'N_pyr_y': 3,
+                   'tstop': 25,
+                   })
+    # initialise an unconnected network
+    net = Network(params)
+
+    # issue _weak_ excitatory drive to distal apical dendrites
+    # NB must not cause Na- or Ca-spiking, as these are not associated with
+    # dipolar fields
+    weights_nmda = {'L2_basket': .0, 'L2_pyramidal': .0005,
+                    'L2_basket': .0, 'L5_pyramidal': .0005}
+    net.add_evoked_drive('d', mu=10., sigma=0., numspikes=1, location='distal',
+                         sync_within_trial=True, weights_nmda=weights_nmda)
+
+    conductivity = 0.3
+
+    # create far-field grid of LFP electrodes; note that cells are assumed
+    # to lie in the XZ-plane
+    xmin, xmax = -5e4, 5e4
+    zmin, zmax = -5e4, 5e4
+    step = 5e3
+    posy = 1e2  # out-of-plane
+    electrode_pos = list()
+    for posx in np.arange(xmin, xmax, step):
+        for posz in np.arange(zmin, zmax, step):
+            electrode_pos.append((posx, posy, posz))
+    net.add_electrode_array('grid_psa', electrode_pos,
+                            conductivity=conductivity, method='psa')
+    net.add_electrode_array('grid_lsa', electrode_pos,
+                            conductivity=conductivity, method='lsa')
+
+    with MPIBackend(n_procs=2):
+        dpl = simulate_dipole(net, postproc=False)
+
+    X_p = np.arange(xmin, xmax, step) / 1000
+    Z_p = np.arange(zmin, zmax, step) / 1000
+    Y_p = posy / 1000
+    idt = np.argmin(np.abs(dpl[0].times - 15.))
+    phi_p_psa = np.zeros((len(X_p), len(Z_p)))
+    phi_p_lsa = np.zeros((len(X_p), len(Z_p)))
+    phi_p_theory = np.zeros((len(X_p), len(Z_p)))
+
+    # location of equivalent current dipole for this stimulation (manual)
+    # XXX this still assumes apical dendrites aligned with Y in Neuron
```

Comment:
Is this comment still relevant after #353 ?

Reply:
Nope! Removing!